File: php/src/UseCase/OrderCreationUseCase.php
<?php
declare(strict_types=1);

namespace Pitchart\TellDontAskKata\UseCase;

use Pitchart\TellDontAskKata\Domain\Order;
use Pitchart\TellDontAskKata\Domain\OrderItem;
use Pitchart\TellDontAskKata\Domain\OrderStatus;
use Pitchart\TellDontAskKata\Repository\OrderRepository;
use Pitchart\TellDontAskKata\Repository\ProductCatalog;

class OrderCreationUseCase
{
    private OrderRepository $repository;

    private ProductCatalog $catalog;

    /**
     * @param OrderRepository $repository
     * @param ProductCatalog $catalog
     */
    public function __construct(OrderRepository $repository, ProductCatalog $catalog)
    {
        $this->repository = $repository;
        $this->catalog = $catalog;
    }

    /**
     * @throws UnknownProductException
     */
    public function run(SellItemsRequest $request): void
    {
        $order = (new Order())
            ->setStatus(OrderStatus::Created)
            ->setCurrency("EUR");

        /** @var SellItemRequest $itemRequest */
        foreach ($request->getItems() as $itemRequest) {
            $product = $this->catalog->getByName($itemRequest->getProductName());

            if ($product == null) {
                throw new UnknownProductException();
            }

            $unitaryTax = self::round(($product->getPrice() / 100) * $product->getCategory()->getTaxPercentage());
            $unitaryTaxedAmount = self::round($product->getPrice() + $unitaryTax);
            $taxedAmount = self::round($unitaryTaxedAmount * $itemRequest->getQuantity());
            $taxAmount = self::round($unitaryTax * $itemRequest->getQuantity());

            $orderItem = (new OrderItem())
                ->setProduct($product)
                ->setQuantity($itemRequest->getQuantity())
                ->setTax($taxAmount)
                ->setTaxedAmount($taxedAmount);

            $order->getItems()->add($orderItem);
            $order->setTotal($order->getTotal() + $taxedAmount);
            $order->setTax($order->getTax() + $taxAmount);

            $this->repository->save($order);
        }

    }

    private static function round(float $amount): float
    {
        return round($amount, 2);
    }
}

File: csharp/TellDontAsk/UseCase/OrderCreationUseCase.cs
using Deveel.Math;
using System.Collections.Generic;
using TellDontAsk.Domain;
using TellDontAsk.Repository;

//import static java.math.BigDecimal.valueOf;
//import static java.math.RoundingMode.HALF_UP;

namespace TellDontAsk.UseCase
{
    public class OrderCreationUseCase
    {
        private readonly OrderRepository orderRepository;
        private readonly ProductCatalog productCatalog;

        public OrderCreationUseCase(OrderRepository orderRepository, ProductCatalog productCatalog)
        {
            this.orderRepository = orderRepository;
            this.productCatalog = productCatalog;
        }

        public void run(SellItemsRequest request)
        {
            Order order = new Order();
            order.setStatus(OrderStatus.CREATED);
            order.setItems(new List<OrderItem>());
            order.setCurrency("EUR");
            order.setTotal(new BigDecimal(0.0));
            order.setTax(new BigDecimal(0.0));

            foreach (SellItemRequest itemRequest in request.getRequests())
            {
                Product product = productCatalog.getByName(itemRequest.getProductName());

                if (product == null)
                {
                    throw new UnknownProductException();
                }
                else
                {
                    BigDecimal unitaryTax = product.getPrice().Divide(BigDecimal.ValueOf(100)).Multiply(product.getCategory().getTaxPercentage()).SetScale(2, RoundingMode.HalfUp);
                    BigDecimal unitaryTaxedAmount = product.getPrice().Add(unitaryTax).SetScale(2, RoundingMode.HalfUp);
                    BigDecimal taxedAmount = unitaryTaxedAmount.Multiply(BigDecimal.ValueOf(itemRequest.getQuantity())).SetScale(2, RoundingMode.HalfUp);
                    BigDecimal taxAmount = unitaryTax.Multiply(BigDecimal.ValueOf(itemRequest.getQuantity()));

                    OrderItem orderItem = new OrderItem();
                    orderItem.setProduct(product);
                    orderItem.setQuantity(itemRequest.getQuantity());
                    orderItem.setTax(taxAmount);
                    orderItem.setTaxedAmount(taxedAmount);
                    order.getItems().Add(orderItem);

                    order.setTotal(order.getTotal().Add(taxedAmount));
                    order.setTax(order.getTax().Add(taxAmount));
                }
            }

            orderRepository.save(order);
        }
    }
}
